{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-16 17:17:17.990360: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-06-16 17:17:19.229529: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory\n",
      "2024-06-16 17:17:19.229623: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory\n",
      "2024-06-16 17:17:19.229632: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "3\n",
      "26\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-16 17:17:21.405789: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open shared object file: No such file or directory\n",
      "2024-06-16 17:17:21.405850: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1934] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.\n",
      "Skipping registering GPU devices...\n"
     ]
    }
   ],
   "source": [
    "import hugging_face.tiger as hf\n",
    "print(hf.CONTEXT_3P)\n",
    "print(hf.CONTEXT_5P)\n",
    "print(hf.TARGET_LEN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# project-wide nucleotide tokens\n",
    "NUCLEOTIDE_TOKENS = dict(zip(['A', 'C', 'G', 'T'], [0, 1, 2, 3]))\n",
    "\n",
    "# relevant columns\n",
    "INDEX_COLS = ['gene', 'guide_id']\n",
    "LFC_COLS = ['lfc_r1', 'lfc_r2', 'lfc_r3']\n",
    "SEQUENCE_FEATS = ['guide_seq', 'target_seq', '5p_context', '3p_context']\n",
    "FEATURE_GROUPS = {\n",
    "    'target location': [\n",
    "        'loc_utr_5p',\n",
    "        'loc_cds',\n",
    "        'loc_utr_3p',\n",
    "        'log_gene_len',\n",
    "        # 'strand'\n",
    "    ],\n",
    "    'junction overlap': [\n",
    "        'junction_olap_5p',\n",
    "        'junction_olap_3p',\n",
    "    ],\n",
    "    'junction proximity': [\n",
    "        'junction_dist_5p',\n",
    "        'junction_dist_3p',\n",
    "    ],\n",
    "    'guide secondary structure': [\n",
    "        'direct_repeat',\n",
    "        'g_quad',\n",
    "    ],\n",
    "    'guide MFE': [\n",
    "        'mfe',\n",
    "    ],\n",
    "    'hybridization MFE': [\n",
    "        'hybrid_mfe_1_23',\n",
    "        'hybrid_mfe_15_9',\n",
    "        'hybrid_mfe_3_12',\n",
    "    ],\n",
    "    'target accessibility': [\n",
    "        'log_unpaired',\n",
    "        'log_unpaired_11',\n",
    "        'log_unpaired_19',\n",
    "        'log_unpaired_25',\n",
    "    ],\n",
    "    # 'localization': [\n",
    "    #     'perc_gene_nuc',\n",
    "    #     'perc_junc_nuc',\n",
    "    # ],\n",
    "    # 'usage': [\n",
    "    #     'perc_junc_use',\n",
    "    # ],\n",
    "}\n",
    "SCALAR_FEATS = [feature for group_features in FEATURE_GROUPS.values() for feature in group_features]\n",
    "SCALAR_FEATS_INDELS = set(SCALAR_FEATS) - {'hybrid_mfe_1_23', 'hybrid_mfe_15_9', 'hybrid_mfe_3_12'}\n",
    "UNIT_SCALED_FEATS = ['loc_utr_5p', 'loc_cds', 'loc_utr_3p',\n",
    "                     'direct_repeat', 'g_quad',\n",
    "                     'perc_gene_nuc', 'perc_junc_nuc', 'perc_junc_use']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_data(dataset, pm_only=False, indels=False, holdout='targets', scale_non_seq_feats=False):\n",
    "    \"\"\"\n",
    "    Loads specified dataset and its corresponding non-targeting data (if it exists)\n",
    "    :param dataset: which dataset to use\n",
    "    :param pm_only: only include perfectly matched guides\n",
    "    :param indels: whether to include guides with insertions or deletions\n",
    "    :param holdout: what should be held out in each fold {genes, guides, targets}\n",
    "    :param scale_non_seq_feats: whether to scale non-sequence features to unit interval\n",
    "    :return: two DataFrames containing targeting and non-targeting data, respectively\n",
    "    \"\"\"\n",
    "    data_file = os.path.join('data-processed', dataset + '.bz2')\n",
    "    assert os.path.exists(data_file), 'if assertion fails, run: python data.py'\n",
    "\n",
    "    # load data\n",
    "    data = pd.read_pickle(data_file)\n",
    "\n",
    "    # filtered data as requested\n",
    "    if pm_only:\n",
    "        data = data[data.guide_type == 'PM']\n",
    "    elif not indels:\n",
    "        data = data[data.guide_type.isin(['PM', 'SM', 'DM', 'RDM', 'TM', 'RTM'])]\n",
    "\n",
    "    # set the folds\n",
    "    if holdout == 'genes':\n",
    "        data['fold'] = data.get('gene')\n",
    "    elif holdout == 'guides':\n",
    "        data['fold'] = data.get('guide_fold')\n",
    "    elif holdout == 'targets':\n",
    "        data['fold'] = data.get('target_fold')\n",
    "    else:\n",
    "        raise NotImplementedError\n",
    "\n",
    "    # load non-targeting data if it exists\n",
    "    nt_file = os.path.join('data-processed', dataset + '-nt.bz2')\n",
    "    data_nt = pd.read_pickle(nt_file) if os.path.exists(nt_file) else None\n",
    "\n",
    "    # scalar feature transformation and scaling\n",
    "    for feature in SCALAR_FEATS:\n",
    "        if feature in data.columns:\n",
    "            if 'junction_dist' in feature:\n",
    "                data[feature] = data[feature].apply(lambda x: np.log10(1 + np.abs(x)))\n",
    "            if scale_non_seq_feats and feature not in UNIT_SCALED_FEATS:\n",
    "                data[feature] -= data[feature].min()\n",
    "                scale = data[feature].max()\n",
    "                if scale > 0:\n",
    "                    data[feature] /= scale\n",
    "                print(feature, data[feature].min(), data[feature].max())\n",
    "\n",
    "    return data, data_nt\n",
    "\n",
    "\n",
    "def label_and_filter_data(data, data_nt, nt_quantile=0.01, method='MinActiveRatio', min_active_ratio=0.1, quiet=True):\n",
    "    \"\"\"\n",
    "    Labels guide activity as active (LFC < specified non-targeting quantile) and filters our non-essential genes\n",
    "    :param data: screen data as a DataFrame\n",
    "    :param data_nt: non-targeting data as a DataFrame\n",
    "    :param nt_quantile: non-targeting quantile that defines the threshold under which guides are considered active\n",
    "    :param method: essential gene filter method\n",
    "    :param min_active_ratio: used by MinActiveRatio--genes with an active guide ratio less than this value get removed\n",
    "    :param quiet: silences Lilliefors non-targeting distribution tests\n",
    "    :return: filtered data with observed labels (active vs inactive)\n",
    "    \"\"\"\n",
    "    # non-targeting data is available\n",
    "    if data_nt is not None and len(data_nt) > 0:\n",
    "\n",
    "        # compute mean of replicates\n",
    "        data_nt['lfc'] = data_nt[LFC_COLS].mean(axis=1)\n",
    "\n",
    "        # set active threshold based on quantile of non-targeting distribution (assumed to be normal)\n",
    "        threshold = norm.ppf(q=nt_quantile, loc=data_nt['lfc'].mean(), scale=data_nt['lfc'].std())\n",
    "        if not quiet:\n",
    "            _, p_val = lilliefors(data_nt['lfc'].values)\n",
    "            print('Lilliefors p-value of NT replicate medians: {:.4e}'.format(p_val))\n",
    "            print('A {:.4f} quantile yields an LFC threshold of {:.4f}'.format(nt_quantile, threshold))\n",
    "\n",
    "    # non-targeting data is unavailable, so use default threshold\n",
    "    else:\n",
    "        threshold = -0.5\n",
    "\n",
    "    # provided target values\n",
    "    if set(LFC_COLS).issubset(data.columns):\n",
    "\n",
    "        # take mean of replicates as target value\n",
    "        data['observed_lfc'] = data[LFC_COLS].mean(axis=1)\n",
    "        data = data[~data['observed_lfc'].isna()]\n",
    "        assert sum(np.isnan(data['observed_lfc'].values)) == 0\n",
    "\n",
    "    # label guides as active/inactive\n",
    "    if 'observed_lfc' in data.columns:\n",
    "        data['observed_label'] = data['observed_lfc'] < threshold\n",
    "    elif 'q' in data.columns:\n",
    "        data['observed_label'] = (data['q'] < 0.1) & (data['z'] < 0.)  # significant at FDR 0.1 in the correct direction\n",
    "\n",
    "    # keep only single-unique and common junctions\n",
    "    if 'junction_category' in data.columns:\n",
    "        data = data.loc[data['junction_category'].isin(['single_unique', 'common'])]\n",
    "\n",
    "    # apply filter\n",
    "    if method == 'NoFilter':\n",
    "        return data\n",
    "    elif method == 'MinActiveRatio' and 'observed_label' in data.columns:\n",
    "        df = pd.DataFrame(data[data.guide_type == 'PM'].groupby('gene')['observed_label'].mean().rename('active ratio'))\n",
    "        return data[data['gene'].isin(df[df['active ratio'] >= min_active_ratio].index.values)]\n",
    "    else:\n",
    "        raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "# import utils\n",
    "import pandas as pd\n",
    "# # import tensorflow as tf\n",
    "# from data import load_data, label_and_filter_data, model_inputs, SCALAR_FEATS\n",
    "# from models import build_model, train_model, test_model\n",
    "# from normalization import get_normalization_object\n",
    "\n",
    "# load, label, and filter data\n",
    "data = load_data('off-target')#, pm_only=args.pm_only, indels=args.indels, holdout=args.holdout)\n",
    "# data = label_and_filter_data(*data, args.nt_quantile, args.filter_method, args.min_active_ratio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import norm\n",
    "data = label_and_filter_data(*data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>guide_id</th>\n",
       "      <th>guide_type</th>\n",
       "      <th>lfc_r1</th>\n",
       "      <th>lfc_r2</th>\n",
       "      <th>lfc_r3</th>\n",
       "      <th>guide_seq</th>\n",
       "      <th>target_seq</th>\n",
       "      <th>5p_context</th>\n",
       "      <th>3p_context</th>\n",
       "      <th>...</th>\n",
       "      <th>hybrid_mfe_3_12</th>\n",
       "      <th>log_unpaired</th>\n",
       "      <th>log_unpaired_11</th>\n",
       "      <th>log_unpaired_19</th>\n",
       "      <th>log_unpaired_25</th>\n",
       "      <th>guide_fold</th>\n",
       "      <th>target_fold</th>\n",
       "      <th>fold</th>\n",
       "      <th>observed_lfc</th>\n",
       "      <th>observed_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371</td>\n",
       "      <td>PM</td>\n",
       "      <td>-3.819780</td>\n",
       "      <td>-3.044753</td>\n",
       "      <td>-1.709841</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTACCA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.8</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-2.858125</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P1-A:G</td>\n",
       "      <td>SM</td>\n",
       "      <td>-3.445606</td>\n",
       "      <td>-3.799614</td>\n",
       "      <td>-3.011046</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTACCG</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.8</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-3.418755</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P2-C:T</td>\n",
       "      <td>SM</td>\n",
       "      <td>-2.723421</td>\n",
       "      <td>-2.168753</td>\n",
       "      <td>-3.163318</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTACTA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.8</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-2.685164</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P3-C:A</td>\n",
       "      <td>SM</td>\n",
       "      <td>-1.698941</td>\n",
       "      <td>-2.151166</td>\n",
       "      <td>-0.916911</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTAACA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-18.7</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-1.589006</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P4-A:T</td>\n",
       "      <td>SM</td>\n",
       "      <td>-2.435223</td>\n",
       "      <td>-3.483880</td>\n",
       "      <td>-2.727288</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTTCCA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-17.4</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-2.882130</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118262</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA1067:1086-1108</td>\n",
       "      <td>PM</td>\n",
       "      <td>-0.037639</td>\n",
       "      <td>0.239060</td>\n",
       "      <td>-0.230432</td>\n",
       "      <td>ACTCCTCCTAGACTTCTCCTGTT</td>\n",
       "      <td>TGAGGAGGATCTGAAGAGGACAA</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>TGATGGCCTGTGGAGGCTCAATCCAGACCAGTGTGAATGCTCTGTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>-24.4</td>\n",
       "      <td>-2.975745</td>\n",
       "      <td>-0.574501</td>\n",
       "      <td>-0.210624</td>\n",
       "      <td>-1.639831</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.009671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118263</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA1576:1632-1654</td>\n",
       "      <td>PM</td>\n",
       "      <td>-1.877323</td>\n",
       "      <td>-2.398506</td>\n",
       "      <td>-2.439100</td>\n",
       "      <td>CCACGCCTAGTTACGCGACTGTC</td>\n",
       "      <td>GGTGCGGATCAATGCGCTGACAG</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>CAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAACCAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-24.2</td>\n",
       "      <td>-6.231938</td>\n",
       "      <td>-0.663498</td>\n",
       "      <td>-1.973189</td>\n",
       "      <td>-0.726269</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>-2.238310</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118264</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA0299:299-321</td>\n",
       "      <td>PM</td>\n",
       "      <td>-0.839628</td>\n",
       "      <td>-1.834822</td>\n",
       "      <td>-0.205472</td>\n",
       "      <td>CTCCGTTTCGTTGTTAAAGATTA</td>\n",
       "      <td>GAGGCAAAGCAACAATTTCTAAT</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>GATGGGGCCACAATTCTGAAACTTCTTGATGTTGTCCATCCTGCAG...</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.9</td>\n",
       "      <td>-1.870503</td>\n",
       "      <td>-0.402233</td>\n",
       "      <td>-1.353552</td>\n",
       "      <td>-1.252402</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.959974</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118265</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA1573:1629-1651</td>\n",
       "      <td>PM</td>\n",
       "      <td>-2.477016</td>\n",
       "      <td>-4.210653</td>\n",
       "      <td>-1.269312</td>\n",
       "      <td>ATACCACGCCTAGTTACGCGACT</td>\n",
       "      <td>TATGGTGCGGATCAATGCGCTGA</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>CAGCAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAAC...</td>\n",
       "      <td>...</td>\n",
       "      <td>-23.4</td>\n",
       "      <td>-5.529942</td>\n",
       "      <td>-0.156943</td>\n",
       "      <td>-0.841205</td>\n",
       "      <td>-0.854693</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>-2.652327</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118266</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA0143:143-165</td>\n",
       "      <td>PM</td>\n",
       "      <td>-1.636634</td>\n",
       "      <td>-1.388903</td>\n",
       "      <td>-1.447765</td>\n",
       "      <td>ACTACGGGTGTGGTCAATAGGAT</td>\n",
       "      <td>TGATGCCCACACCAGTTATCCTA</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>TTGAAAGAGGGGACTGATAGCTCCCAAGGCATCCCCCAGCTTGTGA...</td>\n",
       "      <td>...</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-2.238866</td>\n",
       "      <td>-0.671673</td>\n",
       "      <td>-0.481528</td>\n",
       "      <td>-0.049977</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-1.491101</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93145 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene                     guide_id guide_type    lfc_r1    lfc_r2  \\\n",
       "0       SNRNP200            crRNA0349:349-371         PM -3.819780 -3.044753   \n",
       "1       SNRNP200  crRNA0349:349-371_SM_P1-A:G         SM -3.445606 -3.799614   \n",
       "2       SNRNP200  crRNA0349:349-371_SM_P2-C:T         SM -2.723421 -2.168753   \n",
       "3       SNRNP200  crRNA0349:349-371_SM_P3-C:A         SM -1.698941 -2.151166   \n",
       "4       SNRNP200  crRNA0349:349-371_SM_P4-A:T         SM -2.435223 -3.483880   \n",
       "...          ...                          ...        ...       ...       ...   \n",
       "118262      CCT7          crRNA1067:1086-1108         PM -0.037639  0.239060   \n",
       "118263      CCT7          crRNA1576:1632-1654         PM -1.877323 -2.398506   \n",
       "118264      CCT7            crRNA0299:299-321         PM -0.839628 -1.834822   \n",
       "118265      CCT7          crRNA1573:1629-1651         PM -2.477016 -4.210653   \n",
       "118266      CCT7            crRNA0143:143-165         PM -1.636634 -1.388903   \n",
       "\n",
       "          lfc_r3                guide_seq               target_seq  \\\n",
       "0      -1.709841  AGCCTCCCGTAACTACTCTACCA  TCGGAGGGCATTGATGAGATGGT   \n",
       "1      -3.011046  AGCCTCCCGTAACTACTCTACCG  TCGGAGGGCATTGATGAGATGGT   \n",
       "2      -3.163318  AGCCTCCCGTAACTACTCTACTA  TCGGAGGGCATTGATGAGATGGT   \n",
       "3      -0.916911  AGCCTCCCGTAACTACTCTAACA  TCGGAGGGCATTGATGAGATGGT   \n",
       "4      -2.727288  AGCCTCCCGTAACTACTCTTCCA  TCGGAGGGCATTGATGAGATGGT   \n",
       "...          ...                      ...                      ...   \n",
       "118262 -0.230432  ACTCCTCCTAGACTTCTCCTGTT  TGAGGAGGATCTGAAGAGGACAA   \n",
       "118263 -2.439100  CCACGCCTAGTTACGCGACTGTC  GGTGCGGATCAATGCGCTGACAG   \n",
       "118264 -0.205472  CTCCGTTTCGTTGTTAAAGATTA  GAGGCAAAGCAACAATTTCTAAT   \n",
       "118265 -1.269312  ATACCACGCCTAGTTACGCGACT  TATGGTGCGGATCAATGCGCTGA   \n",
       "118266 -1.447765  ACTACGGGTGTGGTCAATAGGAT  TGATGCCCACACCAGTTATCCTA   \n",
       "\n",
       "                                               5p_context  \\\n",
       "0       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "1       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "2       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "3       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "4       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "...                                                   ...   \n",
       "118262  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118263  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118264  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118265  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118266  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "\n",
       "                                               3p_context  ...  \\\n",
       "0       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "1       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "2       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "3       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "4       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "...                                                   ...  ...   \n",
       "118262  TGATGGCCTGTGGAGGCTCAATCCAGACCAGTGTGAATGCTCTGTC...  ...   \n",
       "118263  CAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAACCAT...  ...   \n",
       "118264  GATGGGGCCACAATTCTGAAACTTCTTGATGTTGTCCATCCTGCAG...  ...   \n",
       "118265  CAGCAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAAC...  ...   \n",
       "118266  TTGAAAGAGGGGACTGATAGCTCCCAAGGCATCCCCCAGCTTGTGA...  ...   \n",
       "\n",
       "        hybrid_mfe_3_12  log_unpaired  log_unpaired_11  log_unpaired_19  \\\n",
       "0                 -19.8     -4.052203        -0.760030        -0.557244   \n",
       "1                 -19.8     -4.052203        -0.760030        -0.557244   \n",
       "2                 -19.8     -4.052203        -0.760030        -0.557244   \n",
       "3                 -18.7     -4.052203        -0.760030        -0.557244   \n",
       "4                 -17.4     -4.052203        -0.760030        -0.557244   \n",
       "...                 ...           ...              ...              ...   \n",
       "118262            -24.4     -2.975745        -0.574501        -0.210624   \n",
       "118263            -24.2     -6.231938        -0.663498        -1.973189   \n",
       "118264            -16.9     -1.870503        -0.402233        -1.353552   \n",
       "118265            -23.4     -5.529942        -0.156943        -0.841205   \n",
       "118266            -23.0     -2.238866        -0.671673        -0.481528   \n",
       "\n",
       "        log_unpaired_25  guide_fold  target_fold  fold  observed_lfc  \\\n",
       "0             -0.797968           3            9     9     -2.858125   \n",
       "1             -0.797968           5            9     9     -3.418755   \n",
       "2             -0.797968           2            9     9     -2.685164   \n",
       "3             -0.797968           3            9     9     -1.589006   \n",
       "4             -0.797968          10            9     9     -2.882130   \n",
       "...                 ...         ...          ...   ...           ...   \n",
       "118262        -1.639831           3            6     6     -0.009671   \n",
       "118263        -0.726269           6            6     6     -2.238310   \n",
       "118264        -1.252402           2            4     4     -0.959974   \n",
       "118265        -0.854693           6            7     7     -2.652327   \n",
       "118266        -0.049977           1            9     9     -1.491101   \n",
       "\n",
       "        observed_label  \n",
       "0                 True  \n",
       "1                 True  \n",
       "2                 True  \n",
       "3                 True  \n",
       "4                 True  \n",
       "...                ...  \n",
       "118262           False  \n",
       "118263            True  \n",
       "118264            True  \n",
       "118265            True  \n",
       "118266            True  \n",
       "\n",
       "[93145 rows x 31 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalizer = get_normalization_object('No')(data)\n",
    "data = normalizer.normalize_targets(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_normalization_object(method: str):\n",
    "    if method == 'No':\n",
    "        return NoNormalization\n",
    "    elif method == 'FrequentistQuantile':\n",
    "        return FrequentistQuantileNormalization\n",
    "    elif method == 'UnitInterval':\n",
    "        return UnitIntervalNormalization\n",
    "    elif method == 'UnitVariance':\n",
    "        return UnitVarianceNormalization\n",
    "    elif method == 'UnitMeanOfSquares':\n",
    "        return UnitMeanOfSquaresNormalization\n",
    "    elif method == 'ZeroMeanUnitVariance':\n",
    "        return ZeroMeanUnitVarianceNormalization\n",
    "    elif method == 'DepletionRatio':\n",
    "        return DepletionRatioNormalization\n",
    "    elif method == 'QuantileMatching':\n",
    "        return QuantileMatchingNormalization\n",
    "    else:\n",
    "        raise NotImplementedError\n",
    "\n",
    "\n",
    "class Normalization(object):\n",
    "    def __init__(self, df: pd.DataFrame):\n",
    "        self.original_lfc = df[['gene', 'guide_seq', 'observed_lfc']].copy().set_index(['gene', 'guide_seq'])\n",
    "        assert not self.original_lfc.index.has_duplicates\n",
    "\n",
    "    def normalize_targets(self, df: pd.DataFrame):\n",
    "        raise NotImplementedError\n",
    "\n",
    "    def denormalize_observations(self, df: pd.DataFrame):\n",
    "        # restore observed values\n",
    "        if 'observed_lfc' in df.columns:\n",
    "            df = pd.merge(\n",
    "                left=df,\n",
    "                right=self.original_lfc,\n",
    "                left_on=['gene', 'guide_seq'],\n",
    "                right_index=True,\n",
    "                suffixes=('_normalized', '')\n",
    "            )\n",
    "        if 'observed_pm_lfc' in df.columns:\n",
    "            df_pm = df.loc[df.guide_type == 'PM', ['gene', 'target_seq', 'observed_lfc']]\n",
    "            df_pm = df_pm.rename(columns={'observed_lfc': 'observed_pm_lfc'}).set_index(['gene', 'target_seq'])\n",
    "            df = pd.merge(\n",
    "                left=df,\n",
    "                right=df_pm,\n",
    "                how='left',\n",
    "                left_on=['gene', 'target_seq'],\n",
    "                right_index=True,\n",
    "                suffixes=('_normalized', '')\n",
    "            )\n",
    "        return df\n",
    "\n",
    "    def denormalize_predictions(self, df: pd.DataFrame):\n",
    "        df['predicted_lfc_normalized'] = df['predicted_lfc']\n",
    "        if 'predicted_pm_lfc' in df.columns:\n",
    "            df['predicted_pm_lfc_normalized'] = df['predicted_pm_lfc']\n",
    "        return df\n",
    "\n",
    "    def denormalize_targets_and_predictions(self, df: pd.DataFrame):\n",
    "        return self.denormalize_predictions(self.denormalize_observations(df.copy()))\n",
    "\n",
    "\n",
    "class NoNormalization(Normalization):\n",
    "    def __init__(self, df: pd.DataFrame):\n",
    "        super().__init__(df)\n",
    "        self.output_fn = 'linear'\n",
    "\n",
    "    def normalize_targets(self, df: pd.DataFrame):\n",
    "        return df\n",
    "\n",
    "    def denormalize_observations(self, df: pd.DataFrame):\n",
    "        return df\n",
    "\n",
    "    def denormalize_predictions(self, df: pd.DataFrame):\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Fix for server\n",
    "# export CUDNN_PATH=\"/home/amirali/dtsui/anaconda3/envs/tiger/lib/python3.10/site-packages/nvidia/cudnn\"\n",
    "# export LD_LIBRARY_PATH=\"$CUDNN_PATH/lib\":\"/usr/local/cuda/lib64\"\n",
    "# # ...\n",
    "# export PATH=\"$PATH\":\"/usr/local/cuda/bin\"\n",
    "# export TF_CPP_MIN_LOG_LEVEL=\"2\"\n",
    "\"\"\"\n",
    "import sys\n",
    "# print(sys.path)\n",
    "path_to_remove = '/nethome/dtsui31/.local/lib/python3.10/site-packages'\n",
    "if path_to_remove in sys.path:\n",
    "    sys.path.remove(path_to_remove)\n",
    "import tensorflow as tf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-04 16:23:12.449634: W tensorflow/tsl/framework/cpu_allocator_impl.cc:82] Allocation of 4435340568 exceeds 10% of free system memory.\n",
      "2024-06-04 16:23:26.143513: W tensorflow/tsl/framework/cpu_allocator_impl.cc:82] Allocation of 6425274192 exceeds 10% of free system memory.\n"
     ]
    }
   ],
   "source": [
    "target_seq = tf.stack([tf.constant(list(seq)) for seq in data['target_seq']], axis=0)\n",
    "left_context = tf.ragged.stack([tf.constant(list(seq), tf.string) for seq in data['5p_context']], axis=0)\n",
    "right_context = tf.ragged.stack([tf.constant(list(seq), tf.string) for seq in data['3p_context']], axis=0)\n",
    "guide_seq = tf.ragged.stack([tf.constant(list(seq)) for seq in data['guide_seq']], axis=0)\n",
    "\n",
    "nucleotide_table = tf.lookup.StaticVocabularyTable(\n",
    "    initializer=tf.lookup.KeyValueTensorInitializer(\n",
    "        keys=tf.constant(list(NUCLEOTIDE_TOKENS.keys()) + ['N'], dtype=tf.string),\n",
    "        values=tf.constant(list(NUCLEOTIDE_TOKENS.values()) + [255], dtype=tf.int64)),\n",
    "    num_oov_buckets=1)\n",
    "target_tokens = nucleotide_table.lookup(target_seq)\n",
    "tokens_5p = nucleotide_table.lookup(left_context)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "context = 5\n",
    "if isinstance(context, (list, tuple)):\n",
    "    context_5p, context_3p = tuple(context)\n",
    "else:\n",
    "    context_5p = context_3p = context\n",
    "\n",
    "# add target context\n",
    "max_len_5p = max(tokens_5p.bounding_shape()[1], context_5p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.preprocessing.sequence import pad_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.preprocessing.sequence import pad_sequences\n",
    "\n",
    "def model_inputs(data, context, *, scalar_feats=(), target_feats=(), include_replicates=False, max_context=100):\n",
    "    \"\"\"\n",
    "    Prepares a dictionary of model inputs and target values from the provided DataFrame\n",
    "    :param data: panda's DataFrame containing model inputs and target values\n",
    "    :param context: amount of target context\n",
    "    :param scalar_feats: scalar features to be provided to the model\n",
    "    :param target_feats: additional sequence features\n",
    "    :param include_replicates: whether to include raw replicates\n",
    "    :param max_context: maximum amount of up- and down-stream context (reduce RAM usage for guides with 1kb available)\n",
    "    :return: dictionary containing model inputs and target values\n",
    "    \"\"\"\n",
    "    # shuffle rows (tensorflow shuffling is approximate, doing a full shuffle here makes it more exact)\n",
    "    data = data.sample(frac=1)\n",
    "\n",
    "    # keep only data without NaN values for the scalar features\n",
    "    data = data[(~data[list(scalar_feats)].isna()).product(1) == 1]\n",
    "\n",
    "    # trim context to some reasonable amount\n",
    "    data['5p_context'] = data['5p_context'].apply(lambda s: s[-min(len(s), max_context):]).astype(str)\n",
    "    data['3p_context'] = data['3p_context'].apply(lambda s: s[:min(len(s), max_context)]).astype(str)\n",
    "\n",
    "    # load target and guide sequences\n",
    "    target_seq = tf.stack([tf.constant(list(seq)) for seq in data['target_seq']], axis=0)\n",
    "    left_context = tf.ragged.stack([tf.constant(list(seq), tf.string) for seq in data['5p_context']], axis=0)\n",
    "    right_context = tf.ragged.stack([tf.constant(list(seq), tf.string) for seq in data['3p_context']], axis=0)\n",
    "    guide_seq = tf.ragged.stack([tf.constant(list(seq)) for seq in data['guide_seq']], axis=0)\n",
    "\n",
    "    # convert nucleotides to integer codes\n",
    "    nucleotide_table = tf.lookup.StaticVocabularyTable(\n",
    "        initializer=tf.lookup.KeyValueTensorInitializer(\n",
    "            keys=tf.constant(list(NUCLEOTIDE_TOKENS.keys()) + ['N'], dtype=tf.string),\n",
    "            values=tf.constant(list(NUCLEOTIDE_TOKENS.values()) + [255], dtype=tf.int64)),\n",
    "        num_oov_buckets=1)\n",
    "    target_tokens = nucleotide_table.lookup(target_seq)\n",
    "    tokens_5p = nucleotide_table.lookup(left_context)\n",
    "    tokens_3p = nucleotide_table.lookup(right_context)\n",
    "    guide_tokens = tf.RaggedTensor.from_row_splits(values=nucleotide_table.lookup(guide_seq.values),\n",
    "                                                   row_splits=guide_seq.row_splits).to_tensor(255)\n",
    "    target_tokens = tf.cast(target_tokens, tf.uint8)\n",
    "    guide_tokens = tf.cast(guide_tokens, tf.uint8)\n",
    "\n",
    "    # these operations are only necessary if using additional target context sequence\n",
    "    if isinstance(context, (list, tuple)):\n",
    "        context_5p, context_3p = tuple(context)\n",
    "    else:\n",
    "        context_5p = context_3p = context\n",
    "\n",
    "    # add target context\n",
    "    max_len_5p = max(tokens_5p.bounding_shape()[1], context_5p)\n",
    "    tokens_5p = pad_sequences(tokens_5p.to_list(), maxlen=max_len_5p, dtype='uint8', padding='pre', value=255)\n",
    "    tokens_5p = tf.constant(tokens_5p[:, tokens_5p.shape[1]-context_5p:tokens_5p.shape[1]])\n",
    "    max_len_3p = max(tokens_3p.bounding_shape()[1], context_3p)\n",
    "    tokens_3p = pad_sequences(tokens_3p.to_list(), maxlen=max_len_3p, dtype='uint8', padding='post', value=255)\n",
    "    tokens_3p = tf.constant(tokens_3p[:, :context_3p])\n",
    "\n",
    "    # assemble dictionary of core model inputs\n",
    "    inputs = {\n",
    "        # data identifiers for downstream analysis\n",
    "        'gene': tf.constant(data['gene'], tf.string),\n",
    "        'target_seq': tf.constant(data['target_seq'], tf.string),\n",
    "        'guide_id': tf.constant(data['guide_id'], tf.string),\n",
    "        'guide_seq': tf.constant(data['guide_seq'], tf.string),\n",
    "        'guide_type': tf.constant(data['guide_type'], tf.string),\n",
    "        # sequence features\n",
    "        'target_tokens': target_tokens,\n",
    "        'guide_tokens': guide_tokens,\n",
    "        '5p_tokens': tokens_5p,\n",
    "        '3p_tokens': tokens_3p,\n",
    "    }\n",
    "\n",
    "    # target values\n",
    "    if 'observed_lfc' in data.columns:\n",
    "        inputs.update({'observed_lfc': tf.constant(data['observed_lfc'], tf.float32)})\n",
    "    if 'observed_label' in data.columns:\n",
    "        inputs.update({'observed_label': tf.constant(data['observed_label'], tf.uint8)})\n",
    "    if 'sample_weights' in data.columns:\n",
    "        inputs.update({'sample_weights': tf.constant(data['sample_weights'], tf.float32)})\n",
    "    if include_replicates:\n",
    "        inputs.update({'replicate_lfc': tf.constant(data[LFC_COLS], tf.float32)})\n",
    "\n",
    "    # add sequence features\n",
    "    features = tf.zeros(target_tokens.shape[:1] + [context_5p + target_tokens.shape[1] + context_3p] + [0])\n",
    "    for feature in target_feats:\n",
    "        if feature in data.columns:\n",
    "            feature = tf.constant(np.stack(data[feature]), tf.float32)\n",
    "            features = tf.concat([features, tf.expand_dims(feature, axis=-1)], axis=-1)\n",
    "        else:\n",
    "            raise Exception('Missing target feature: ' + feature)\n",
    "    inputs.update({'target_features': features})\n",
    "\n",
    "    # add optional features\n",
    "    for feature in scalar_feats:\n",
    "        if feature in data.columns:\n",
    "            inputs.update({feature: tf.constant(data[feature], tf.float32)})\n",
    "        else:\n",
    "            raise Exception('Missing scalar feature: ' + feature)\n",
    "\n",
    "    return inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_data = model_inputs(data, context=(2,0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<tf.Tensor: shape=(93145, 23), dtype=uint8, numpy=\n",
       "array([[3, 2, 0, ..., 0, 0, 2],\n",
       "       [3, 1, 2, ..., 0, 3, 1],\n",
       "       [1, 0, 2, ..., 3, 1, 3],\n",
       "       ...,\n",
       "       [2, 0, 3, ..., 1, 0, 3],\n",
       "       [2, 2, 2, ..., 2, 3, 3],\n",
       "       [2, 0, 1, ..., 3, 0, 1]], dtype=uint8)>"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data['target_tokens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Union"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SequenceModelWithNonSequenceFeatures(object):\n",
    "    def __init__(self):\n",
    "        self.non_sequence_features = None\n",
    "\n",
    "    def concatenate_non_sequence_features(self, data, x, scalar_feats):\n",
    "        non_sequence_features = []\n",
    "        for feature in (scalar_feats if self.non_sequence_features is None else self.non_sequence_features):\n",
    "            if feature in data.keys():\n",
    "                non_sequence_features.append(feature)\n",
    "                x = tf.concat([x, tf.cast(data[feature][:, None], tf.float32)], axis=1)\n",
    "\n",
    "        if self.non_sequence_features is None:\n",
    "            self.non_sequence_features = non_sequence_features\n",
    "        else:\n",
    "            assert set(self.non_sequence_features) == set(non_sequence_features)\n",
    "\n",
    "        return x\n",
    "\n",
    "\n",
    "class TargetSequenceWithRBP(SequenceModelWithNonSequenceFeatures):\n",
    "    def __init__(self, guide_len: int, context_5p: int, context_3p: int, *, rbp_list: list, **kwargs):\n",
    "        super().__init__()\n",
    "        self.input_parser = layers.TargetSequenceAndPositionalFeatures(guide_len, context_5p, context_3p, len(rbp_list))\n",
    "        self.rbp_list = rbp_list\n",
    "\n",
    "        # model declaration\n",
    "        self.model = tf.keras.Sequential(name='TargetSequenceWithRBP', layers=[\n",
    "            layers.SequenceSequentialWithNonSequenceBypass(\n",
    "                input_parser=self.input_parser,\n",
    "                sequence_layers=[\n",
    "                    layers.ReduceAndConcatTargetRBP(self.input_parser.feature_channels),\n",
    "                    tf.keras.layers.Conv1D(filters=64, kernel_size=4, activation='relu', padding='same'),\n",
    "                    tf.keras.layers.Conv1D(filters=64, kernel_size=4, activation='relu', padding='same'),\n",
    "                    tf.keras.layers.MaxPool1D(pool_size=2, padding='same'),\n",
    "                    tf.keras.layers.Flatten(),\n",
    "                    tf.keras.layers.Dropout(0.25),\n",
    "                ]),\n",
    "            tf.keras.layers.Dense(units=128, activation='sigmoid'),\n",
    "            tf.keras.layers.Dropout(0.1),\n",
    "            tf.keras.layers.Dense(units=32, activation='sigmoid'),\n",
    "            tf.keras.layers.Dropout(0.1),\n",
    "            tf.keras.layers.Dense(1, activation='linear')\n",
    "        ])\n",
    "\n",
    "    def pack_inputs(self, data: dict, scalar_feats: Union[list, tuple] = tuple(SCALAR_FEATS)):\n",
    "\n",
    "        # one-hot encode, flatten, and concatenate target (and context) sequence tokens\n",
    "        x = tf.concat([data['5p_tokens'], data['target_tokens'], data['3p_tokens']], axis=1)\n",
    "        x = tf.reshape(tf.one_hot(x, depth=4), [len(data['target_tokens']), -1])\n",
    "\n",
    "        # if we are using additional non-sequence but positional target features\n",
    "        if self.input_parser.feature_channels > 0:\n",
    "            x = tf.concat([x, tf.reshape(data['target_features'], [len(data['target_tokens']), -1])], axis=1)\n",
    "\n",
    "        # concatenate and log available non-sequence features\n",
    "        x = self.concatenate_non_sequence_features(data, x, scalar_feats)\n",
    "\n",
    "        # target values\n",
    "        y = data['observed_lfc'] if 'observed_lfc' in data.keys() else None\n",
    "\n",
    "        return x, y\n",
    "\n",
    "class OneHotSequenceModel(SequenceModelWithNonSequenceFeatures):\n",
    "    def __init__(self, target_len: int, context_5p: int, context_3p: int, use_guide_seq: bool, pad_guide_seq: bool):\n",
    "        super().__init__()\n",
    "        self.input_parser = layers.OneHotInputParser(target_len, context_5p, context_3p, use_guide_seq, pad_guide_seq)\n",
    "\n",
    "    def pack_inputs(self, data: dict, scalar_feats: Union[list, tuple] = tuple(SCALAR_FEATS)):\n",
    "\n",
    "        # one-hot encode, flatten, and concatenate target (and context) sequence tokens\n",
    "        x = tf.concat([data['5p_tokens'], data['target_tokens'], data['3p_tokens']], axis=1)\n",
    "        x = tf.reshape(tf.one_hot(x, depth=4), [len(data['target_tokens']), -1])\n",
    "\n",
    "        # if we are using guide sequence, do the same to those tokens (but pad them first to match target + context)\n",
    "        if self.input_parser.guide_len > 0:\n",
    "\n",
    "            # pre- and post-pad guides with zero to match combined target and context sequence\n",
    "            data_type = data['guide_tokens'].dtype\n",
    "            pad_5p = 255 * tf.ones([data['guide_tokens'].shape[0], data['5p_tokens'].shape[1]], dtype=data_type)\n",
    "            pad_3p = 255 * tf.ones([data['guide_tokens'].shape[0], data['3p_tokens'].shape[1]], dtype=data_type)\n",
    "            guide_tokens = tf.concat([pad_5p, data['guide_tokens'], pad_3p], axis=1)\n",
    "            x = tf.concat([x, tf.reshape(tf.one_hot(guide_tokens, depth=4), [len(data['guide_tokens']), -1])], axis=1)\n",
    "\n",
    "        # concatenate and log available non-sequence features\n",
    "        x = self.concatenate_non_sequence_features(data, x, scalar_feats)\n",
    "\n",
    "        # target values\n",
    "        y = data['observed_lfc'] if 'observed_lfc' in data.keys() else None\n",
    "        w = data['sample_weights'] if 'sample_weights' in data.keys() else None\n",
    "\n",
    "        return x, y, w\n",
    "\n",
    "    def parse_input_scores(self, scores):\n",
    "\n",
    "        # unpack scores\n",
    "        target_scores, guide_scores, non_sequence_scores = self.input_parser.call(scores)\n",
    "\n",
    "        # load scores into DataFrame\n",
    "        score_dict = dict()\n",
    "        for nt, token in NUCLEOTIDE_TOKENS.items():\n",
    "            score_dict.update({'target:' + nt: target_scores[..., token].numpy().tolist()})\n",
    "            score_dict.update({'guide:' + nt: guide_scores[..., token].numpy().tolist()})\n",
    "        for i, feature in enumerate(self.non_sequence_features):\n",
    "            score_dict.update({feature: non_sequence_scores[:, i]})\n",
    "        df = pd.DataFrame(score_dict)\n",
    "\n",
    "        return df\n",
    "\n",
    "import layers\n",
    "\n",
    "class Tiger1D(OneHotSequenceModel):\n",
    "    def __init__(self, target_len: int, context_5p: int, context_3p: int, use_guide_seq: bool, **kwargs):\n",
    "        OneHotSequenceModel.__init__(self, target_len, context_5p, context_3p, use_guide_seq, pad_guide_seq=True)\n",
    "\n",
    "        self.model = tf.keras.Sequential(name='Tiger1D', layers=[\n",
    "            layers.SequenceSequentialWithNonSequenceBypass(\n",
    "                input_parser=self.input_parser,\n",
    "                sequence_layers=[\n",
    "                    layers.AlignOneHotEncoding1D(use_guide_seq),\n",
    "                    tf.keras.layers.Conv1D(filters=64, kernel_size=4, activation='relu', padding='same'),\n",
    "                    tf.keras.layers.Conv1D(filters=64, kernel_size=4, activation='relu', padding='same'),\n",
    "                    tf.keras.layers.MaxPool1D(pool_size=2, padding='same'),\n",
    "                    tf.keras.layers.Flatten(),\n",
    "                    tf.keras.layers.Dropout(0.25),\n",
    "                ]),\n",
    "            tf.keras.layers.Dense(units=128, activation='sigmoid'),\n",
    "            tf.keras.layers.Dropout(0.1),\n",
    "            tf.keras.layers.Dense(units=32, activation='sigmoid'),\n",
    "            tf.keras.layers.Dropout(0.1),\n",
    "            tf.keras.layers.Dense(1, activation=kwargs.get('output_fn') or 'linear')\n",
    "        ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def common_parser_arguments():\n",
    "\n",
    "    # common arguments\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('--batch_size', type=int, default=2048, help='tensorflow batch size')\n",
    "    parser.add_argument('--context', type=str, default=None, help='amount of target sequence context')\n",
    "    parser.add_argument('--dataset', type=str, default='off-target', help='which dataset to use')\n",
    "    parser.add_argument('--debug', action='store_true', default=False, help='debug mode will run models eagerly')\n",
    "    parser.add_argument('--fig_ext', type=str, default='.pdf', help='which file extension to use when saving plots')\n",
    "    parser.add_argument('--filter_method', type=str, default=None, help='gene filtering method')\n",
    "    parser.add_argument('--holdout', type=str, default='targets', help='how to assemble cross-validation folds')\n",
    "    parser.add_argument('--indels', action='store_true', default=False, help='include targets with indels')\n",
    "    parser.add_argument('--kwargs', type=str, default=None, help='model hyper-parameters')\n",
    "    parser.add_argument('--loss', type=str, default='log_cosh', help='training loss function')\n",
    "    parser.add_argument('--min_active_ratio', type=float, default=None, help='ratio of active guides to keep a gene')\n",
    "    parser.add_argument('--model', type=str, default=None, help='model name')\n",
    "    parser.add_argument('--normalization', type=str, default=None, help='normalization method')\n",
    "    parser.add_argument('--normalization_kwargs', type=str, default=None, help='normalization parameters')\n",
    "    parser.add_argument('--nt_quantile', type=float, default=None, help='active guide non-targeting quantile threshold')\n",
    "    parser.add_argument('--pm_only', action='store_true', default=False, help='use only perfect match guides')\n",
    "    parser.add_argument('--use_guide_seq', action='store_true', default=False, help='use guide sequence for PM only')\n",
    "    parser.add_argument('--seed', type=int, default=None, help='random number seed')\n",
    "    parser.add_argument('--seq_only', action='store_true', default=False, help='sequence only model')\n",
    "\n",
    "    return parser\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "build_model() missing 1 required positional argument: 'loss_fn'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[57], line 31\u001b[0m\n\u001b[1;32m     27\u001b[0m     model\u001b[38;5;241m.\u001b[39mmodel\u001b[38;5;241m.\u001b[39mcompile(optimizer\u001b[38;5;241m=\u001b[39moptimizer, loss\u001b[38;5;241m=\u001b[39mloss_fn, weighted_metrics\u001b[38;5;241m=\u001b[39m[], run_eagerly\u001b[38;5;241m=\u001b[39mdebug)\n\u001b[1;32m     29\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m model\n\u001b[0;32m---> 31\u001b[0m model \u001b[38;5;241m=\u001b[39m \u001b[43mbuild_model\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtest\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     32\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mtarget_len\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrain_data\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtarget_tokens\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshape\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     33\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mcontext_5p\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrain_data\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m5p_tokens\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshape\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     34\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mcontext_3p\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrain_data\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m3p_tokens\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshape\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     35\u001b[0m \u001b[43m                    \u001b[49m\u001b[43muse_guide_seq\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     36\u001b[0m \u001b[43m                    \u001b[49m\u001b[38;5;66;43;03m# loss_fn=args.loss,\u001b[39;49;00m\n\u001b[1;32m     37\u001b[0m \u001b[43m                    \u001b[49m\u001b[38;5;66;43;03m# debug=args.debug,\u001b[39;49;00m\n\u001b[1;32m     38\u001b[0m \u001b[43m                    \u001b[49m\u001b[38;5;66;43;03m# output_fn=normalizer.output_fn,\u001b[39;49;00m\n\u001b[1;32m     39\u001b[0m \u001b[43m)\u001b[49m\n\u001b[1;32m     40\u001b[0m \u001b[38;5;66;03m# model = train_model(model, train_data, train_data, args.batch_size)\u001b[39;00m\n\u001b[1;32m     41\u001b[0m \n\u001b[1;32m     42\u001b[0m \u001b[38;5;66;03m# # accumulate targets and predictions for held-out fold\u001b[39;00m\n\u001b[1;32m     43\u001b[0m \u001b[38;5;66;03m# predictions = pd.concat([predictions, test_model(model, valid_data)])\u001b[39;00m\n",
      "\u001b[0;31mTypeError\u001b[0m: build_model() missing 1 required positional argument: 'loss_fn'"
     ]
    }
   ],
   "source": [
    "from models import build_model, train_model, test_model\n",
    "import utils\n",
    "\n",
    "# script arguments\n",
    "parser = utils.common_parser_arguments()\n",
    "# parser.add_argument('--mm_only', action='store_true', default=False, help='use only mismatched guides')\n",
    "# args = utils.parse_common_arguments(parser)\n",
    "# assert not (args.mm_only and args.pm_only)\n",
    "\n",
    "# train mode\n",
    "\n",
    "def build_model(name, target_len, context_5p, context_3p, use_guide_seq, loss_fn, debug=False, **kwargs):\n",
    "    if name == 'Tiger1D':\n",
    "        model = Tiger1D(target_len, context_5p, context_3p, use_guide_seq, **kwargs)\n",
    "        optimizer = tf.optimizers.Adam(1e-3)\n",
    "    elif name == 'Tiger2D':\n",
    "        model = Tiger2D(target_len, context_5p, context_3p, use_guide_seq, **kwargs)\n",
    "        optimizer = tf.optimizers.Adam(1e-3)\n",
    "    elif name == 'TargetSequenceWithRBP':\n",
    "        model = TargetSequenceWithRBP(target_len, context_5p, context_3p, **kwargs)\n",
    "        optimizer = tf.optimizers.Adam(1e-3)\n",
    "    # elif name == 'TranscriptTransformer':\n",
    "    #     model = TranscriptTransformer(target_len, guide_len, use_guide_seq, **kwargs)\n",
    "    #     optimizer = tf.optimizers.Adam(5e-4)\n",
    "    else:\n",
    "        raise NotImplementedError\n",
    "    model.model.compile(optimizer=optimizer, loss=loss_fn, weighted_metrics=[], run_eagerly=debug)\n",
    "\n",
    "    return model\n",
    "\n",
    "model = build_model(name='test',\n",
    "                    target_len=train_data['target_tokens'].shape[1],\n",
    "                    context_5p=train_data['5p_tokens'].shape[1],\n",
    "                    context_3p=train_data['3p_tokens'].shape[1],\n",
    "                    use_guide_seq=False,\n",
    "                    loss_fn='log_cosh',\n",
    "                    # debug=args.debug,\n",
    "                    # output_fn=normalizer.output_fn,\n",
    ")\n",
    "# model = train_model(model, train_data, train_data, args.batch_size)\n",
    "\n",
    "# # accumulate targets and predictions for held-out fold\n",
    "# predictions = pd.concat([predictions, test_model(model, valid_data)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = build_model(name='Tiger1D',\n",
    "    target_len=train_data['target_tokens'].shape[1],\n",
    "    context_5p=train_data['5p_tokens'].shape[1],\n",
    "    context_3p=train_data['3p_tokens'].shape[1],\n",
    "    use_guide_seq=False,\n",
    "    loss_fn='log_cosh',\n",
    "    debug=False,\n",
    "    output_fn=normalizer.output_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictions = test_model(model, train_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_seq</th>\n",
       "      <th>gene</th>\n",
       "      <th>guide_id</th>\n",
       "      <th>guide_seq</th>\n",
       "      <th>guide_type</th>\n",
       "      <th>observed_lfc</th>\n",
       "      <th>observed_label</th>\n",
       "      <th>predicted_lfc</th>\n",
       "      <th>observed_pm_lfc</th>\n",
       "      <th>predicted_pm_lfc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAACAAACCTTTATCAGGATGA</td>\n",
       "      <td>PSMD1</td>\n",
       "      <td>crRNA1573:1654-1676</td>\n",
       "      <td>TTTTGTTTGGAAATAGTCCTACT</td>\n",
       "      <td>PM</td>\n",
       "      <td>0.070935</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.255048</td>\n",
       "      <td>0.070935</td>\n",
       "      <td>-0.255048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAACAACGGGAAGATTGAACTC</td>\n",
       "      <td>EIF3B</td>\n",
       "      <td>crRNA1748:1840-1862</td>\n",
       "      <td>TTTTGTTGCCCTTCTAACTTGAG</td>\n",
       "      <td>PM</td>\n",
       "      <td>0.134618</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.277627</td>\n",
       "      <td>0.134618</td>\n",
       "      <td>-0.277627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAACAAGCTCCGGCTGCATTAC</td>\n",
       "      <td>NUP93</td>\n",
       "      <td>crRNA1224:1224-1246</td>\n",
       "      <td>TTTTGTTCGAGGCCGACGTAATG</td>\n",
       "      <td>PM</td>\n",
       "      <td>0.330254</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.271856</td>\n",
       "      <td>0.330254</td>\n",
       "      <td>-0.271856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAACAATTTTACATGTGTAGCA</td>\n",
       "      <td>WDR75</td>\n",
       "      <td>crRNA0605:668-690</td>\n",
       "      <td>TTTTGTTAAAATGTACACATCGT</td>\n",
       "      <td>PM</td>\n",
       "      <td>0.077236</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.274825</td>\n",
       "      <td>0.077236</td>\n",
       "      <td>-0.274825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAACAGAGATTCTTCCTCCCTT</td>\n",
       "      <td>SF3B1</td>\n",
       "      <td>crRNA2346:2511-2533</td>\n",
       "      <td>TTTTGTCTCTAAGAAGGAGGGAA</td>\n",
       "      <td>PM</td>\n",
       "      <td>-0.230626</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.253106</td>\n",
       "      <td>-0.230626</td>\n",
       "      <td>-0.253106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93140</th>\n",
       "      <td>TTTTTGATGAAAATGCTGATGAG</td>\n",
       "      <td>BUB1B</td>\n",
       "      <td>crRNA0978:1016-1038</td>\n",
       "      <td>AAAAACTACTTTTACGACTACTC</td>\n",
       "      <td>PM</td>\n",
       "      <td>0.375851</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.266358</td>\n",
       "      <td>0.375851</td>\n",
       "      <td>-0.266358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93141</th>\n",
       "      <td>TTTTTGATGATGACCTTGAAGAT</td>\n",
       "      <td>POLA1</td>\n",
       "      <td>crRNA0301:342-364</td>\n",
       "      <td>AAAAACTACTACTGGAACTTCTA</td>\n",
       "      <td>PM</td>\n",
       "      <td>0.021497</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.261367</td>\n",
       "      <td>0.021497</td>\n",
       "      <td>-0.261367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93142</th>\n",
       "      <td>TTTTTGCAGCTGTTATCAATATC</td>\n",
       "      <td>BUB1B</td>\n",
       "      <td>crRNA2602:2658-2680</td>\n",
       "      <td>AAAAACGTCGACAATAGTTATAG</td>\n",
       "      <td>PM</td>\n",
       "      <td>0.030771</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.262369</td>\n",
       "      <td>0.030771</td>\n",
       "      <td>-0.262369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93143</th>\n",
       "      <td>TTTTTGGAATTTTATCTCCTAGT</td>\n",
       "      <td>NUP133</td>\n",
       "      <td>crRNA0877:877-899</td>\n",
       "      <td>AAAAACCTTAAAATAGAGGATCA</td>\n",
       "      <td>PM</td>\n",
       "      <td>0.184384</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.260188</td>\n",
       "      <td>0.184384</td>\n",
       "      <td>-0.260188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93144</th>\n",
       "      <td>TTTTTGTCATCTGTAATCCAGAA</td>\n",
       "      <td>PSMD1</td>\n",
       "      <td>crRNA0918:940-962</td>\n",
       "      <td>AAAAACAGTAGACATTAGGTCTT</td>\n",
       "      <td>PM</td>\n",
       "      <td>0.184956</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.268237</td>\n",
       "      <td>0.184956</td>\n",
       "      <td>-0.268237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93145 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    target_seq    gene             guide_id  \\\n",
       "0      AAAACAAACCTTTATCAGGATGA   PSMD1  crRNA1573:1654-1676   \n",
       "1      AAAACAACGGGAAGATTGAACTC   EIF3B  crRNA1748:1840-1862   \n",
       "2      AAAACAAGCTCCGGCTGCATTAC   NUP93  crRNA1224:1224-1246   \n",
       "3      AAAACAATTTTACATGTGTAGCA   WDR75    crRNA0605:668-690   \n",
       "4      AAAACAGAGATTCTTCCTCCCTT   SF3B1  crRNA2346:2511-2533   \n",
       "...                        ...     ...                  ...   \n",
       "93140  TTTTTGATGAAAATGCTGATGAG   BUB1B  crRNA0978:1016-1038   \n",
       "93141  TTTTTGATGATGACCTTGAAGAT   POLA1    crRNA0301:342-364   \n",
       "93142  TTTTTGCAGCTGTTATCAATATC   BUB1B  crRNA2602:2658-2680   \n",
       "93143  TTTTTGGAATTTTATCTCCTAGT  NUP133    crRNA0877:877-899   \n",
       "93144  TTTTTGTCATCTGTAATCCAGAA   PSMD1    crRNA0918:940-962   \n",
       "\n",
       "                     guide_seq guide_type  observed_lfc  observed_label  \\\n",
       "0      TTTTGTTTGGAAATAGTCCTACT         PM      0.070935               0   \n",
       "1      TTTTGTTGCCCTTCTAACTTGAG         PM      0.134618               0   \n",
       "2      TTTTGTTCGAGGCCGACGTAATG         PM      0.330254               0   \n",
       "3      TTTTGTTAAAATGTACACATCGT         PM      0.077236               0   \n",
       "4      TTTTGTCTCTAAGAAGGAGGGAA         PM     -0.230626               0   \n",
       "...                        ...        ...           ...             ...   \n",
       "93140  AAAAACTACTTTTACGACTACTC         PM      0.375851               0   \n",
       "93141  AAAAACTACTACTGGAACTTCTA         PM      0.021497               0   \n",
       "93142  AAAAACGTCGACAATAGTTATAG         PM      0.030771               0   \n",
       "93143  AAAAACCTTAAAATAGAGGATCA         PM      0.184384               0   \n",
       "93144  AAAAACAGTAGACATTAGGTCTT         PM      0.184956               0   \n",
       "\n",
       "       predicted_lfc  observed_pm_lfc  predicted_pm_lfc  \n",
       "0          -0.255048         0.070935         -0.255048  \n",
       "1          -0.277627         0.134618         -0.277627  \n",
       "2          -0.271856         0.330254         -0.271856  \n",
       "3          -0.274825         0.077236         -0.274825  \n",
       "4          -0.253106        -0.230626         -0.253106  \n",
       "...              ...              ...               ...  \n",
       "93140      -0.266358         0.375851         -0.266358  \n",
       "93141      -0.261367         0.021497         -0.261367  \n",
       "93142      -0.262369         0.030771         -0.262369  \n",
       "93143      -0.260188         0.184384         -0.260188  \n",
       "93144      -0.268237         0.184956         -0.268237  \n",
       "\n",
       "[93145 rows x 10 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<tf.Tensor: shape=(93145, 2), dtype=uint8, numpy=\n",
       "array([[2, 0],\n",
       "       [1, 0],\n",
       "       [0, 1],\n",
       "       ...,\n",
       "       [1, 1],\n",
       "       [2, 0],\n",
       "       [1, 1]], dtype=uint8)>"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data['5p_tokens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "x, _, _ = model.pack_inputs(train_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene': <tf.Tensor: shape=(93145,), dtype=string, numpy=\n",
       " array([b'SF3B1', b'AQR', b'NUP133', ..., b'EIF3B', b'NUP133', b'SF3B1'],\n",
       "       dtype=object)>,\n",
       " 'target_seq': <tf.Tensor: shape=(93145,), dtype=string, numpy=\n",
       " array([b'TGAACCGCTATTGATTGATGAAG', b'TCGTGCAGATGTTACCATAAATC',\n",
       "        b'CAGGGACAAATATAACTTTTTCT', ..., b'GATGACCCTCAGGCCATAATCAT',\n",
       "        b'GGGAGGTGATAAGACTTACAGTT', b'GACGATGACTATTCATCATCTAC'],\n",
       "       dtype=object)>,\n",
       " 'guide_id': <tf.Tensor: shape=(93145,), dtype=string, numpy=\n",
       " array([b'crRNA1682:1826-1848_RDM_P14-T:A|P20-T:C',\n",
       "        b'crRNA1700:1831-1853_SM_P9-G:T',\n",
       "        b'crRNA3473:3549-3571_RDM_P8-G:T|P19-C:A', ...,\n",
       "        b'crRNA1178:1251-1273_TM_P1-A:T|P2-T:A|P3-G:A',\n",
       "        b'crRNA0701:701-723_RTM_P2-A:C|P12-T:C|P16-C:T',\n",
       "        b'crRNA0291:291-313_RDM_P14-A:T|P21-G:T'], dtype=object)>,\n",
       " 'guide_seq': <tf.Tensor: shape=(93145,), dtype=string, numpy=\n",
       " array([b'ACTCGGCGAAAACTAACTACTTC', b'AGCACGTCTACAATTGTATTTAG',\n",
       "        b'GTCCATGTTTATATTTAAAAAGA', ..., b'CTACTGGGAGTCCGGTATTAAAT',\n",
       "        b'CCCTCCATTATCCTGAATGTCCA', b'CTTCTACTGTTAAGTAGTAGATG'],\n",
       "       dtype=object)>,\n",
       " 'guide_type': <tf.Tensor: shape=(93145,), dtype=string, numpy=array([b'RDM', b'SM', b'RDM', ..., b'TM', b'RTM', b'RDM'], dtype=object)>,\n",
       " 'target_tokens': <tf.Tensor: shape=(93145, 23), dtype=uint8, numpy=\n",
       " array([[3, 2, 0, ..., 0, 0, 2],\n",
       "        [3, 1, 2, ..., 0, 3, 1],\n",
       "        [1, 0, 2, ..., 3, 1, 3],\n",
       "        ...,\n",
       "        [2, 0, 3, ..., 1, 0, 3],\n",
       "        [2, 2, 2, ..., 2, 3, 3],\n",
       "        [2, 0, 1, ..., 3, 0, 1]], dtype=uint8)>,\n",
       " 'guide_tokens': <tf.Tensor: shape=(93145, 23), dtype=uint8, numpy=\n",
       " array([[0, 1, 3, ..., 3, 3, 1],\n",
       "        [0, 2, 1, ..., 3, 0, 2],\n",
       "        [2, 3, 1, ..., 0, 2, 0],\n",
       "        ...,\n",
       "        [1, 3, 0, ..., 0, 0, 3],\n",
       "        [1, 1, 1, ..., 1, 1, 0],\n",
       "        [1, 3, 3, ..., 0, 3, 2]], dtype=uint8)>,\n",
       " '5p_tokens': <tf.Tensor: shape=(93145, 4), dtype=uint8, numpy=\n",
       " array([[3, 1, 0, 3],\n",
       "        [2, 0, 2, 3],\n",
       "        [3, 2, 3, 3],\n",
       "        ...,\n",
       "        [2, 1, 0, 2],\n",
       "        [0, 3, 3, 1],\n",
       "        [3, 2, 0, 3]], dtype=uint8)>,\n",
       " '3p_tokens': <tf.Tensor: shape=(93145, 4), dtype=uint8, numpy=\n",
       " array([[0, 3, 3, 0],\n",
       "        [3, 1, 0, 0],\n",
       "        [0, 0, 0, 0],\n",
       "        ...,\n",
       "        [1, 3, 2, 2],\n",
       "        [3, 1, 1, 3],\n",
       "        [2, 0, 2, 3]], dtype=uint8)>,\n",
       " 'observed_lfc': <tf.Tensor: shape=(93145,), dtype=float32, numpy=\n",
       " array([ 0.52813816, -1.9707322 ,  0.08210623, ...,  0.15911059,\n",
       "         0.22125784, -0.10746585], dtype=float32)>,\n",
       " 'observed_label': <tf.Tensor: shape=(93145,), dtype=uint8, numpy=array([0, 1, 0, ..., 0, 0, 0], dtype=uint8)>,\n",
       " 'target_features': <tf.Tensor: shape=(93145, 31, 0), dtype=float32, numpy=array([], shape=(93145, 31, 0), dtype=float32)>}"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>guide_id</th>\n",
       "      <th>guide_type</th>\n",
       "      <th>lfc_r1</th>\n",
       "      <th>lfc_r2</th>\n",
       "      <th>lfc_r3</th>\n",
       "      <th>guide_seq</th>\n",
       "      <th>target_seq</th>\n",
       "      <th>5p_context</th>\n",
       "      <th>3p_context</th>\n",
       "      <th>...</th>\n",
       "      <th>hybrid_mfe_3_12</th>\n",
       "      <th>log_unpaired</th>\n",
       "      <th>log_unpaired_11</th>\n",
       "      <th>log_unpaired_19</th>\n",
       "      <th>log_unpaired_25</th>\n",
       "      <th>guide_fold</th>\n",
       "      <th>target_fold</th>\n",
       "      <th>fold</th>\n",
       "      <th>observed_lfc</th>\n",
       "      <th>observed_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371</td>\n",
       "      <td>PM</td>\n",
       "      <td>-3.819780</td>\n",
       "      <td>-3.044753</td>\n",
       "      <td>-1.709841</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTACCA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.8</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-2.858125</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P1-A:G</td>\n",
       "      <td>SM</td>\n",
       "      <td>-3.445606</td>\n",
       "      <td>-3.799614</td>\n",
       "      <td>-3.011046</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTACCG</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.8</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-3.418755</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P2-C:T</td>\n",
       "      <td>SM</td>\n",
       "      <td>-2.723421</td>\n",
       "      <td>-2.168753</td>\n",
       "      <td>-3.163318</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTACTA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.8</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-2.685164</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P3-C:A</td>\n",
       "      <td>SM</td>\n",
       "      <td>-1.698941</td>\n",
       "      <td>-2.151166</td>\n",
       "      <td>-0.916911</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTAACA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-18.7</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-1.589006</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P4-A:T</td>\n",
       "      <td>SM</td>\n",
       "      <td>-2.435223</td>\n",
       "      <td>-3.483880</td>\n",
       "      <td>-2.727288</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTTCCA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-17.4</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-2.882130</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118262</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA1067:1086-1108</td>\n",
       "      <td>PM</td>\n",
       "      <td>-0.037639</td>\n",
       "      <td>0.239060</td>\n",
       "      <td>-0.230432</td>\n",
       "      <td>ACTCCTCCTAGACTTCTCCTGTT</td>\n",
       "      <td>TGAGGAGGATCTGAAGAGGACAA</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>TGATGGCCTGTGGAGGCTCAATCCAGACCAGTGTGAATGCTCTGTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>-24.4</td>\n",
       "      <td>-2.975745</td>\n",
       "      <td>-0.574501</td>\n",
       "      <td>-0.210624</td>\n",
       "      <td>-1.639831</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.009671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118263</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA1576:1632-1654</td>\n",
       "      <td>PM</td>\n",
       "      <td>-1.877323</td>\n",
       "      <td>-2.398506</td>\n",
       "      <td>-2.439100</td>\n",
       "      <td>CCACGCCTAGTTACGCGACTGTC</td>\n",
       "      <td>GGTGCGGATCAATGCGCTGACAG</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>CAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAACCAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-24.2</td>\n",
       "      <td>-6.231938</td>\n",
       "      <td>-0.663498</td>\n",
       "      <td>-1.973189</td>\n",
       "      <td>-0.726269</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>-2.238310</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118264</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA0299:299-321</td>\n",
       "      <td>PM</td>\n",
       "      <td>-0.839628</td>\n",
       "      <td>-1.834822</td>\n",
       "      <td>-0.205472</td>\n",
       "      <td>CTCCGTTTCGTTGTTAAAGATTA</td>\n",
       "      <td>GAGGCAAAGCAACAATTTCTAAT</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>GATGGGGCCACAATTCTGAAACTTCTTGATGTTGTCCATCCTGCAG...</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.9</td>\n",
       "      <td>-1.870503</td>\n",
       "      <td>-0.402233</td>\n",
       "      <td>-1.353552</td>\n",
       "      <td>-1.252402</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.959974</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118265</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA1573:1629-1651</td>\n",
       "      <td>PM</td>\n",
       "      <td>-2.477016</td>\n",
       "      <td>-4.210653</td>\n",
       "      <td>-1.269312</td>\n",
       "      <td>ATACCACGCCTAGTTACGCGACT</td>\n",
       "      <td>TATGGTGCGGATCAATGCGCTGA</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>CAGCAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAAC...</td>\n",
       "      <td>...</td>\n",
       "      <td>-23.4</td>\n",
       "      <td>-5.529942</td>\n",
       "      <td>-0.156943</td>\n",
       "      <td>-0.841205</td>\n",
       "      <td>-0.854693</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>-2.652327</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118266</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA0143:143-165</td>\n",
       "      <td>PM</td>\n",
       "      <td>-1.636634</td>\n",
       "      <td>-1.388903</td>\n",
       "      <td>-1.447765</td>\n",
       "      <td>ACTACGGGTGTGGTCAATAGGAT</td>\n",
       "      <td>TGATGCCCACACCAGTTATCCTA</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>TTGAAAGAGGGGACTGATAGCTCCCAAGGCATCCCCCAGCTTGTGA...</td>\n",
       "      <td>...</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-2.238866</td>\n",
       "      <td>-0.671673</td>\n",
       "      <td>-0.481528</td>\n",
       "      <td>-0.049977</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-1.491101</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93145 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene                     guide_id guide_type    lfc_r1    lfc_r2  \\\n",
       "0       SNRNP200            crRNA0349:349-371         PM -3.819780 -3.044753   \n",
       "1       SNRNP200  crRNA0349:349-371_SM_P1-A:G         SM -3.445606 -3.799614   \n",
       "2       SNRNP200  crRNA0349:349-371_SM_P2-C:T         SM -2.723421 -2.168753   \n",
       "3       SNRNP200  crRNA0349:349-371_SM_P3-C:A         SM -1.698941 -2.151166   \n",
       "4       SNRNP200  crRNA0349:349-371_SM_P4-A:T         SM -2.435223 -3.483880   \n",
       "...          ...                          ...        ...       ...       ...   \n",
       "118262      CCT7          crRNA1067:1086-1108         PM -0.037639  0.239060   \n",
       "118263      CCT7          crRNA1576:1632-1654         PM -1.877323 -2.398506   \n",
       "118264      CCT7            crRNA0299:299-321         PM -0.839628 -1.834822   \n",
       "118265      CCT7          crRNA1573:1629-1651         PM -2.477016 -4.210653   \n",
       "118266      CCT7            crRNA0143:143-165         PM -1.636634 -1.388903   \n",
       "\n",
       "          lfc_r3                guide_seq               target_seq  \\\n",
       "0      -1.709841  AGCCTCCCGTAACTACTCTACCA  TCGGAGGGCATTGATGAGATGGT   \n",
       "1      -3.011046  AGCCTCCCGTAACTACTCTACCG  TCGGAGGGCATTGATGAGATGGT   \n",
       "2      -3.163318  AGCCTCCCGTAACTACTCTACTA  TCGGAGGGCATTGATGAGATGGT   \n",
       "3      -0.916911  AGCCTCCCGTAACTACTCTAACA  TCGGAGGGCATTGATGAGATGGT   \n",
       "4      -2.727288  AGCCTCCCGTAACTACTCTTCCA  TCGGAGGGCATTGATGAGATGGT   \n",
       "...          ...                      ...                      ...   \n",
       "118262 -0.230432  ACTCCTCCTAGACTTCTCCTGTT  TGAGGAGGATCTGAAGAGGACAA   \n",
       "118263 -2.439100  CCACGCCTAGTTACGCGACTGTC  GGTGCGGATCAATGCGCTGACAG   \n",
       "118264 -0.205472  CTCCGTTTCGTTGTTAAAGATTA  GAGGCAAAGCAACAATTTCTAAT   \n",
       "118265 -1.269312  ATACCACGCCTAGTTACGCGACT  TATGGTGCGGATCAATGCGCTGA   \n",
       "118266 -1.447765  ACTACGGGTGTGGTCAATAGGAT  TGATGCCCACACCAGTTATCCTA   \n",
       "\n",
       "                                               5p_context  \\\n",
       "0       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "1       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "2       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "3       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "4       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "...                                                   ...   \n",
       "118262  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118263  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118264  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118265  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118266  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "\n",
       "                                               3p_context  ...  \\\n",
       "0       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "1       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "2       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "3       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "4       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "...                                                   ...  ...   \n",
       "118262  TGATGGCCTGTGGAGGCTCAATCCAGACCAGTGTGAATGCTCTGTC...  ...   \n",
       "118263  CAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAACCAT...  ...   \n",
       "118264  GATGGGGCCACAATTCTGAAACTTCTTGATGTTGTCCATCCTGCAG...  ...   \n",
       "118265  CAGCAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAAC...  ...   \n",
       "118266  TTGAAAGAGGGGACTGATAGCTCCCAAGGCATCCCCCAGCTTGTGA...  ...   \n",
       "\n",
       "        hybrid_mfe_3_12  log_unpaired  log_unpaired_11  log_unpaired_19  \\\n",
       "0                 -19.8     -4.052203        -0.760030        -0.557244   \n",
       "1                 -19.8     -4.052203        -0.760030        -0.557244   \n",
       "2                 -19.8     -4.052203        -0.760030        -0.557244   \n",
       "3                 -18.7     -4.052203        -0.760030        -0.557244   \n",
       "4                 -17.4     -4.052203        -0.760030        -0.557244   \n",
       "...                 ...           ...              ...              ...   \n",
       "118262            -24.4     -2.975745        -0.574501        -0.210624   \n",
       "118263            -24.2     -6.231938        -0.663498        -1.973189   \n",
       "118264            -16.9     -1.870503        -0.402233        -1.353552   \n",
       "118265            -23.4     -5.529942        -0.156943        -0.841205   \n",
       "118266            -23.0     -2.238866        -0.671673        -0.481528   \n",
       "\n",
       "        log_unpaired_25  guide_fold  target_fold  fold  observed_lfc  \\\n",
       "0             -0.797968           3            9     9     -2.858125   \n",
       "1             -0.797968           5            9     9     -3.418755   \n",
       "2             -0.797968           2            9     9     -2.685164   \n",
       "3             -0.797968           3            9     9     -1.589006   \n",
       "4             -0.797968          10            9     9     -2.882130   \n",
       "...                 ...         ...          ...   ...           ...   \n",
       "118262        -1.639831           3            6     6     -0.009671   \n",
       "118263        -0.726269           6            6     6     -2.238310   \n",
       "118264        -1.252402           2            4     4     -0.959974   \n",
       "118265        -0.854693           6            7     7     -2.652327   \n",
       "118266        -0.049977           1            9     9     -1.491101   \n",
       "\n",
       "        observed_label  \n",
       "0                 True  \n",
       "1                 True  \n",
       "2                 True  \n",
       "3                 True  \n",
       "4                 True  \n",
       "...                ...  \n",
       "118262           False  \n",
       "118263            True  \n",
       "118264            True  \n",
       "118265            True  \n",
       "118266            True  \n",
       "\n",
       "[93145 rows x 31 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(93145, 31)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_seq = 'AAATCGGAGGGCATTGATGAGATGGTA'\n",
    "TARGET_LEN = 23\n",
    "CONTEXT_5P = 3\n",
    "CONTEXT_3P = 0\n",
    "# nucleotide tokens\n",
    "NUCLEOTIDE_TOKENS = dict(zip(['A', 'C', 'G', 'T', 'N'], [0, 1, 2, 3, 255]))\n",
    "NUCLEOTIDE_COMPLEMENT = dict(zip(['A', 'C', 'G', 'T'], ['T', 'G', 'C', 'A']))\n",
    "\n",
    "def sequence_complement(sequence: list):\n",
    "    return [''.join([NUCLEOTIDE_COMPLEMENT[nt] for nt in list(seq)]) for seq in sequence]\n",
    "\n",
    "def process_data(transcript_seq: str):\n",
    "\n",
    "    # convert to upper case\n",
    "    transcript_seq = transcript_seq.upper()\n",
    "\n",
    "    # get all target sites\n",
    "    target_seq = [transcript_seq[i: i + TARGET_LEN] for i in range(len(transcript_seq) - TARGET_LEN + 1)]\n",
    "\n",
    "    # prepare guide sequences\n",
    "    guide_seq = sequence_complement([seq[CONTEXT_5P:len(seq) - CONTEXT_3P] for seq in target_seq])\n",
    "\n",
    "    # # model inputs\n",
    "    # model_inputs = tf.concat([\n",
    "    #     tf.reshape(one_hot_encode_sequence(target_seq, add_context_padding=False), [len(target_seq), -1]),\n",
    "    #     tf.reshape(one_hot_encode_sequence(guide_seq, add_context_padding=True), [len(guide_seq), -1]),\n",
    "    #     ], axis=-1)\n",
    "    return target_seq, guide_seq#, model_inputs\n",
    "\n",
    "target_seq, guide_seq = process_data(transcript_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AAATCGGAGGGCATTGATGAGAT',\n",
       " 'AATCGGAGGGCATTGATGAGATG',\n",
       " 'ATCGGAGGGCATTGATGAGATGG',\n",
       " 'TCGGAGGGCATTGATGAGATGGT',\n",
       " 'CGGAGGGCATTGATGAGATGGTA']"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CACCGAGCGAAAACATTGGC',\n",
       " 'ACCGAGCGAAAACATTGGCG',\n",
       " 'CCGAGCGAAAACATTGGCGA',\n",
       " 'CGAGCGAAAACATTGGCGAT',\n",
       " 'GAGCGAAAACATTGGCGATC']"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>guide_id</th>\n",
       "      <th>guide_type</th>\n",
       "      <th>lfc_r1</th>\n",
       "      <th>lfc_r2</th>\n",
       "      <th>lfc_r3</th>\n",
       "      <th>guide_seq</th>\n",
       "      <th>target_seq</th>\n",
       "      <th>5p_context</th>\n",
       "      <th>3p_context</th>\n",
       "      <th>...</th>\n",
       "      <th>hybrid_mfe_3_12</th>\n",
       "      <th>log_unpaired</th>\n",
       "      <th>log_unpaired_11</th>\n",
       "      <th>log_unpaired_19</th>\n",
       "      <th>log_unpaired_25</th>\n",
       "      <th>guide_fold</th>\n",
       "      <th>target_fold</th>\n",
       "      <th>fold</th>\n",
       "      <th>observed_lfc</th>\n",
       "      <th>observed_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371</td>\n",
       "      <td>PM</td>\n",
       "      <td>-3.819780</td>\n",
       "      <td>-3.044753</td>\n",
       "      <td>-1.709841</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTACCA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.8</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-2.858125</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P1-A:G</td>\n",
       "      <td>SM</td>\n",
       "      <td>-3.445606</td>\n",
       "      <td>-3.799614</td>\n",
       "      <td>-3.011046</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTACCG</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.8</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-3.418755</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P2-C:T</td>\n",
       "      <td>SM</td>\n",
       "      <td>-2.723421</td>\n",
       "      <td>-2.168753</td>\n",
       "      <td>-3.163318</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTACTA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.8</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-2.685164</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P3-C:A</td>\n",
       "      <td>SM</td>\n",
       "      <td>-1.698941</td>\n",
       "      <td>-2.151166</td>\n",
       "      <td>-0.916911</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTAACA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-18.7</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-1.589006</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNRNP200</td>\n",
       "      <td>crRNA0349:349-371_SM_P4-A:T</td>\n",
       "      <td>SM</td>\n",
       "      <td>-2.435223</td>\n",
       "      <td>-3.483880</td>\n",
       "      <td>-2.727288</td>\n",
       "      <td>AGCCTCCCGTAACTACTCTTCCA</td>\n",
       "      <td>TCGGAGGGCATTGATGAGATGGT</td>\n",
       "      <td>CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...</td>\n",
       "      <td>GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-17.4</td>\n",
       "      <td>-4.052203</td>\n",
       "      <td>-0.760030</td>\n",
       "      <td>-0.557244</td>\n",
       "      <td>-0.797968</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-2.882130</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118262</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA1067:1086-1108</td>\n",
       "      <td>PM</td>\n",
       "      <td>-0.037639</td>\n",
       "      <td>0.239060</td>\n",
       "      <td>-0.230432</td>\n",
       "      <td>ACTCCTCCTAGACTTCTCCTGTT</td>\n",
       "      <td>TGAGGAGGATCTGAAGAGGACAA</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>TGATGGCCTGTGGAGGCTCAATCCAGACCAGTGTGAATGCTCTGTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>-24.4</td>\n",
       "      <td>-2.975745</td>\n",
       "      <td>-0.574501</td>\n",
       "      <td>-0.210624</td>\n",
       "      <td>-1.639831</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.009671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118263</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA1576:1632-1654</td>\n",
       "      <td>PM</td>\n",
       "      <td>-1.877323</td>\n",
       "      <td>-2.398506</td>\n",
       "      <td>-2.439100</td>\n",
       "      <td>CCACGCCTAGTTACGCGACTGTC</td>\n",
       "      <td>GGTGCGGATCAATGCGCTGACAG</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>CAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAACCAT...</td>\n",
       "      <td>...</td>\n",
       "      <td>-24.2</td>\n",
       "      <td>-6.231938</td>\n",
       "      <td>-0.663498</td>\n",
       "      <td>-1.973189</td>\n",
       "      <td>-0.726269</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>-2.238310</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118264</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA0299:299-321</td>\n",
       "      <td>PM</td>\n",
       "      <td>-0.839628</td>\n",
       "      <td>-1.834822</td>\n",
       "      <td>-0.205472</td>\n",
       "      <td>CTCCGTTTCGTTGTTAAAGATTA</td>\n",
       "      <td>GAGGCAAAGCAACAATTTCTAAT</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>GATGGGGCCACAATTCTGAAACTTCTTGATGTTGTCCATCCTGCAG...</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.9</td>\n",
       "      <td>-1.870503</td>\n",
       "      <td>-0.402233</td>\n",
       "      <td>-1.353552</td>\n",
       "      <td>-1.252402</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.959974</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118265</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA1573:1629-1651</td>\n",
       "      <td>PM</td>\n",
       "      <td>-2.477016</td>\n",
       "      <td>-4.210653</td>\n",
       "      <td>-1.269312</td>\n",
       "      <td>ATACCACGCCTAGTTACGCGACT</td>\n",
       "      <td>TATGGTGCGGATCAATGCGCTGA</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>CAGCAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAAC...</td>\n",
       "      <td>...</td>\n",
       "      <td>-23.4</td>\n",
       "      <td>-5.529942</td>\n",
       "      <td>-0.156943</td>\n",
       "      <td>-0.841205</td>\n",
       "      <td>-0.854693</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>-2.652327</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118266</th>\n",
       "      <td>CCT7</td>\n",
       "      <td>crRNA0143:143-165</td>\n",
       "      <td>PM</td>\n",
       "      <td>-1.636634</td>\n",
       "      <td>-1.388903</td>\n",
       "      <td>-1.447765</td>\n",
       "      <td>ACTACGGGTGTGGTCAATAGGAT</td>\n",
       "      <td>TGATGCCCACACCAGTTATCCTA</td>\n",
       "      <td>AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...</td>\n",
       "      <td>TTGAAAGAGGGGACTGATAGCTCCCAAGGCATCCCCCAGCTTGTGA...</td>\n",
       "      <td>...</td>\n",
       "      <td>-23.0</td>\n",
       "      <td>-2.238866</td>\n",
       "      <td>-0.671673</td>\n",
       "      <td>-0.481528</td>\n",
       "      <td>-0.049977</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>-1.491101</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93145 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene                     guide_id guide_type    lfc_r1    lfc_r2  \\\n",
       "0       SNRNP200            crRNA0349:349-371         PM -3.819780 -3.044753   \n",
       "1       SNRNP200  crRNA0349:349-371_SM_P1-A:G         SM -3.445606 -3.799614   \n",
       "2       SNRNP200  crRNA0349:349-371_SM_P2-C:T         SM -2.723421 -2.168753   \n",
       "3       SNRNP200  crRNA0349:349-371_SM_P3-C:A         SM -1.698941 -2.151166   \n",
       "4       SNRNP200  crRNA0349:349-371_SM_P4-A:T         SM -2.435223 -3.483880   \n",
       "...          ...                          ...        ...       ...       ...   \n",
       "118262      CCT7          crRNA1067:1086-1108         PM -0.037639  0.239060   \n",
       "118263      CCT7          crRNA1576:1632-1654         PM -1.877323 -2.398506   \n",
       "118264      CCT7            crRNA0299:299-321         PM -0.839628 -1.834822   \n",
       "118265      CCT7          crRNA1573:1629-1651         PM -2.477016 -4.210653   \n",
       "118266      CCT7            crRNA0143:143-165         PM -1.636634 -1.388903   \n",
       "\n",
       "          lfc_r3                guide_seq               target_seq  \\\n",
       "0      -1.709841  AGCCTCCCGTAACTACTCTACCA  TCGGAGGGCATTGATGAGATGGT   \n",
       "1      -3.011046  AGCCTCCCGTAACTACTCTACCG  TCGGAGGGCATTGATGAGATGGT   \n",
       "2      -3.163318  AGCCTCCCGTAACTACTCTACTA  TCGGAGGGCATTGATGAGATGGT   \n",
       "3      -0.916911  AGCCTCCCGTAACTACTCTAACA  TCGGAGGGCATTGATGAGATGGT   \n",
       "4      -2.727288  AGCCTCCCGTAACTACTCTTCCA  TCGGAGGGCATTGATGAGATGGT   \n",
       "...          ...                      ...                      ...   \n",
       "118262 -0.230432  ACTCCTCCTAGACTTCTCCTGTT  TGAGGAGGATCTGAAGAGGACAA   \n",
       "118263 -2.439100  CCACGCCTAGTTACGCGACTGTC  GGTGCGGATCAATGCGCTGACAG   \n",
       "118264 -0.205472  CTCCGTTTCGTTGTTAAAGATTA  GAGGCAAAGCAACAATTTCTAAT   \n",
       "118265 -1.269312  ATACCACGCCTAGTTACGCGACT  TATGGTGCGGATCAATGCGCTGA   \n",
       "118266 -1.447765  ACTACGGGTGTGGTCAATAGGAT  TGATGCCCACACCAGTTATCCTA   \n",
       "\n",
       "                                               5p_context  \\\n",
       "0       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "1       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "2       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "3       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "4       CGGGAGCAGAGATCTGCGGCCGTTTGCAGCTTGCGGTAGGGAGGCG...   \n",
       "...                                                   ...   \n",
       "118262  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118263  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118264  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118265  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "118266  AGAGTAGCGGAAGTGGTCCGTTCTCTTCCTCTCCCGGCCCAAGCTT...   \n",
       "\n",
       "                                               3p_context  ...  \\\n",
       "0       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "1       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "2       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "3       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "4       GGGCATCATCTACAAGCCCAAAACTAAAGAGACTCGGGAGACCTAT...  ...   \n",
       "...                                                   ...  ...   \n",
       "118262  TGATGGCCTGTGGAGGCTCAATCCAGACCAGTGTGAATGCTCTGTC...  ...   \n",
       "118263  CAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAACCAT...  ...   \n",
       "118264  GATGGGGCCACAATTCTGAAACTTCTTGATGTTGTCCATCCTGCAG...  ...   \n",
       "118265  CAGCAGCCTCTGAGGCTGCGTGCCTGATCGTGTCTGTAGATGAAAC...  ...   \n",
       "118266  TTGAAAGAGGGGACTGATAGCTCCCAAGGCATCCCCCAGCTTGTGA...  ...   \n",
       "\n",
       "        hybrid_mfe_3_12  log_unpaired  log_unpaired_11  log_unpaired_19  \\\n",
       "0                 -19.8     -4.052203        -0.760030        -0.557244   \n",
       "1                 -19.8     -4.052203        -0.760030        -0.557244   \n",
       "2                 -19.8     -4.052203        -0.760030        -0.557244   \n",
       "3                 -18.7     -4.052203        -0.760030        -0.557244   \n",
       "4                 -17.4     -4.052203        -0.760030        -0.557244   \n",
       "...                 ...           ...              ...              ...   \n",
       "118262            -24.4     -2.975745        -0.574501        -0.210624   \n",
       "118263            -24.2     -6.231938        -0.663498        -1.973189   \n",
       "118264            -16.9     -1.870503        -0.402233        -1.353552   \n",
       "118265            -23.4     -5.529942        -0.156943        -0.841205   \n",
       "118266            -23.0     -2.238866        -0.671673        -0.481528   \n",
       "\n",
       "        log_unpaired_25  guide_fold  target_fold  fold  observed_lfc  \\\n",
       "0             -0.797968           3            9     9     -2.858125   \n",
       "1             -0.797968           5            9     9     -3.418755   \n",
       "2             -0.797968           2            9     9     -2.685164   \n",
       "3             -0.797968           3            9     9     -1.589006   \n",
       "4             -0.797968          10            9     9     -2.882130   \n",
       "...                 ...         ...          ...   ...           ...   \n",
       "118262        -1.639831           3            6     6     -0.009671   \n",
       "118263        -0.726269           6            6     6     -2.238310   \n",
       "118264        -1.252402           2            4     4     -0.959974   \n",
       "118265        -0.854693           6            7     7     -2.652327   \n",
       "118266        -0.049977           1            9     9     -1.491101   \n",
       "\n",
       "        observed_label  \n",
       "0                 True  \n",
       "1                 True  \n",
       "2                 True  \n",
       "3                 True  \n",
       "4                 True  \n",
       "...                ...  \n",
       "118262           False  \n",
       "118263            True  \n",
       "118264            True  \n",
       "118265            True  \n",
       "118266            True  \n",
       "\n",
       "[93145 rows x 31 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
